# Data Wrangling project

## Introduction

This project deals with datasets from applicants who applied for positions in various given departments. The applicants are then processed according to their personality scores which will determine the outcome of their applications, based on calculations done in the project. 

## Importing modules

The modules are imported as they are modules that are necessary for completing the calculations and illustrations associated with the project.

In [1]:
import numpy as np
import pandas as pd
from IPython.display import display

## Importing the personality scores dataset 

The CSV file is imported so that the dataset can be used to create a personality data frame.

In [2]:
raw_personality_df = pd.read_csv("../data/personality_scores.csv", sep=";")

The first five entries in the newly created and unedited data frame will be shown.

In [3]:
raw_personality_df.head()

ID Section 5 of 6 [I am always prepared.]  \
0   0                                 (3, 5)   
1   1                                 (3, 5)   
2   2                                 (3, 5)   
3   3                                 (3, 5)   
4   4                                 (3, 3)   

  Section 5 of 6 [I am easily disturbed.]  \
0                                  (4, 5)   
1                                  (4, 5)   
2                                  (4, 3)   
3                                  (4, 5)   
4                                  (4, 5)   

  Section 5 of 6 [I am exacting (demanding) in my work.]  \
0                                             (3, 5)       
1                                             (3, 5)       
2                                             (3, 3)       
3                                             (3, 3)       
4                                             (3, 3)       

  Section 5 of 6 [I am full of ideas.]  \
0                               (5, 5)   
1                               (5, 5)   
2                               (5, 5)   
3                               (5, 5)   
4                               (5, 3)   

  Section 5 of 6 [I am interested in people.]  \
0                                      (2, 3)   
1                                      (2, 5)   
2                                      (2, 5)   
3                                      (2, 5)   
4                                      (2, 3)   

  Section 5 of 6 [I am not interested in abstract ideas.]  \
0                                             (5, 3)        
1                                             (5, 3)        
2                                             (5, 5)        
3                                             (5, 3)        
4                                             (5, 3)        

  Section 5 of 6 [I am not interested in other people's problems.]  \
0                                             (2, 3)                 
1                                             (2, 5)                 
2                                             (2, 5)                 
3                                             (2, 3)                 
4                                             (2, 3)                 

  Section 5 of 6 [I am not really interested in others.]  \
0                                             (2, 5)       
1                                             (2, 5)       
2                                             (2, 5)       
3                                             (2, 3)       
4                                             (2, 3)       

  Section 5 of 6 [I am quick to understand things.]  ... Unnamed: 60  \
0                                            (5, 5)  ...         NaN   
1                                            (5, 5)  ...         NaN   
2                                            (5, 5)  ...         NaN   
3                                            (5, 3)  ...         NaN   
4                                            (5, 5)  ...         NaN   

  Unnamed: 61 Unnamed: 62 Unnamed: 63 Unnamed: 64 Unnamed: 65 Unnamed: 66  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   

  Unnamed: 67 Unnamed: 68 IPIP_HIGH_RISK  
0         NaN         NaN            NaN  
1         NaN         NaN            NaN  
2         NaN         NaN            NaN  
3         NaN         NaN            NaN  
4         NaN         NaN            NaN  

[5 rows x 70 columns]

Information is provided on the newly created data frame's columns.

In [4]:
raw_personality_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1558 entries, 0 to 1557
Data columns (total 70 columns):
 #   Column                                                                    Non-Null Count  Dtype  
---  ------                                                                    --------------  -----  
 0   ID                                                                        1558 non-null   int64  
 1   Section 5 of 6 [I am always prepared.]                                    1558 non-null   object 
 2   Section 5 of 6 [I am easily disturbed.]                                   1558 non-null   object 
 3   Section 5 of 6 [I am exacting (demanding) in my work.]                    1558 non-null   object 
 4   Section 5 of 6 [I am full of ideas.]                                      1558 non-null   object 
 5   Section 5 of 6 [I am interested in people.]                               1558 non-null   object 
 6   Section 5 of 6 [I am not interested in abstract ideas.]         

The information above indicates that there are columns with null entries. Given that the data frame that will be created may not have any null entries, the dataset needs to be cleaned by dropping all the columns in the data frame with empty and null entries. Furthermore, the names of the columns are not satisfactory and will need to be reformatted.

The null and duplicate values and columns are dropped as they could negatively affect the quality and outcome of the analysis. Given that in this project there are individuals who are being critiqued and assessed based on their personality scores, having null and duplicate entries and columns would result in the outcome having skewed, incorrect, or inaccurate data. To avoid this, the rows and columns in the data frame are cleaned, which includes dropping and removing the null and duplicate entries.

The newly created data frame needs to be cleaned so that any duplicate rows, based on the applicant's ID, are removed.

In [5]:
personality_df = raw_personality_df.drop_duplicates(subset=["ID"], keep="last")

All the empty columns will be dropped and cleaned. Furthermore, the names of the columns in the data frame will be reformatted so that they are appropriate to be column names.

In [6]:
personality_df = personality_df.dropna(axis=1)
personality_df.columns = (
    personality_df.columns.str.removeprefix("Section 5 of 6 [")
    .str.removesuffix(".]")
    .str.removesuffix("]")
)
personality_score_df = personality_df.copy()

The first five entries in the updated and cleaned data frame are shown below.

In [7]:
personality_score_df.head()

ID I am always prepared I am easily disturbed  \
0   0               (3, 5)                (4, 5)   
1   1               (3, 5)                (4, 5)   
2   2               (3, 5)                (4, 3)   
3   3               (3, 5)                (4, 5)   
4   4               (3, 3)                (4, 5)   

  I am exacting (demanding) in my work I am full of ideas  \
0                               (3, 5)             (5, 5)   
1                               (3, 5)             (5, 5)   
2                               (3, 3)             (5, 5)   
3                               (3, 3)             (5, 5)   
4                               (3, 3)             (5, 3)   

  I am interested in people I am not interested in abstract ideas  \
0                    (2, 3)                                (5, 3)   
1                    (2, 5)                                (5, 3)   
2                    (2, 5)                                (5, 5)   
3                    (2, 5)                                (5, 3)   
4                    (2, 3)                                (5, 3)   

  I am not interested in other people's problems  \
0                                         (2, 3)   
1                                         (2, 5)   
2                                         (2, 5)   
3                                         (2, 3)   
4                                         (2, 3)   

  I am not really interested in others I am quick to understand things  ...  \
0                               (2, 5)                          (5, 5)  ...   
1                               (2, 5)                          (5, 5)  ...   
2                               (2, 5)                          (5, 5)  ...   
3                               (2, 3)                          (5, 3)  ...   
4                               (2, 3)                          (5, 5)  ...   

  I often forget to put things back in their proper place  \
0                                             (3, 5)        
1                                             (3, 5)        
2                                             (3, 5)        
3                                             (3, 1)        
4                                             (3, 5)        

  I pay attention to details I seldom feel blue (down)  \
0                     (3, 5)                    (4, 3)   
1                     (3, 1)                    (4, 1)   
2                     (3, 5)                    (4, 1)   
3                     (3, 5)                    (4, 1)   
4                     (3, 5)                    (4, 5)   

  I spend time reflecting on things I start conversations  \
0                            (5, 5)                (1, 3)   
1                            (5, 5)                (1, 5)   
2                            (5, 3)                (1, 3)   
3                            (5, 5)                (1, 5)   
4                            (5, 5)                (1, 3)   

  I sympathize with others' feelings I take time out for others  \
0                             (2, 5)                     (2, 5)   
1                             (2, 5)                     (2, 5)   
2                             (2, 5)                     (2, 5)   
3                             (2, 5)                     (2, 5)   
4                             (2, 3)                     (2, 5)   

  I talk to a lot of different people at parties I use difficult words  \
0                                         (1, 3)                (5, 1)   
1                                         (1, 5)                (5, 3)   
2                                         (1, 3)                (5, 1)   
3                                         (1, 5)                (5, 1)   
4                                         (1, 3)                (5, 1)   

  I worry about things  
0               (4, 3)  
1               (4, 3)  
2               (4, 3)  
3               (4, 1)  
4               (4, 3)  

[5 rows x 51 columns]

The data frame above shows that the column names have been reformatted so that they are more appropriate as names for the columns of the data frame.

Information is provided on the cleaned and updated data frame's columns.

In [8]:
personality_score_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1555 entries, 0 to 1557
Data columns (total 51 columns):
 #   Column                                                   Non-Null Count  Dtype 
---  ------                                                   --------------  ----- 
 0   ID                                                       1555 non-null   int64 
 1   I am always prepared                                     1555 non-null   object
 2   I am easily disturbed                                    1555 non-null   object
 3   I am exacting (demanding) in my work                     1555 non-null   object
 4   I am full of ideas                                       1555 non-null   object
 5   I am interested in people                                1555 non-null   object
 6   I am not interested in abstract ideas                    1555 non-null   object
 7   I am not interested in other people's problems           1555 non-null   object
 8   I am not really interested in others  

The information above indicates that the data frame has been cleaned. The columns with null entries have been dropped. This is evidenced by the "Unnamed" columns not being present and also the number of non-null entries decreasing from 1558 non-null entries to 1555 non-null entries.

An assertion is made to determine if the newly created and cleaned data frame has the expected number of rows and columns. The assertion is made to ensure that the newly created and cleaned data frame, which will be used for further calculations, analysis and interpretation in the project,  does not contain any duplicates. This assertion is made by showing that once the previous data frame, called “raw_personality_df”, is cleaned and processed, it will have the same shape as the current, newly created and cleaned data frame called "personality_score_df".

In [9]:
assert (
    len(pd.unique(pd.Series(raw_personality_df["ID"]))),
    len(raw_personality_df.dropna(axis=1).columns),
) == (
    len(personality_score_df),
    len(personality_score_df.columns),
), "The dataframes you are comparing do not have the same shape - they do not have the same number of rows and/or columns"

## Creating new columns from the personality test subscales scores

The data frame will be updated, and new columns will be added. These new columns will contain the total score for each subscale which is calculated from the personality test scores of the respective applicants.

In [10]:
personality_score_totals_df = personality_score_df.copy()

subscale_dictionary = {
    1: "Extraversion",
    2: "Agreeableness",
    3: "Conscientiousness",
    4: "Emotional stability",
    5: "Openness to experience",
}


def calculate_subscale_score(applicant_scores_row, subscale_name):
    subscale_score = 0

    for score in applicant_scores_row[1:]:
        tuple_score = eval(score)
        subscale_score = np.where(
            subscale_dictionary[tuple_score[0]] == subscale_name,
            subscale_score + tuple_score[1],
            subscale_score + 0,
        )

    return subscale_score


for subscale_name_key in subscale_dictionary:
    personality_score_totals_df[subscale_dictionary[subscale_name_key]] = (
        personality_score_df.apply(
            lambda row: calculate_subscale_score(
                row, subscale_dictionary[subscale_name_key]
            ),
            axis=1,
        )
    ).astype("int64")

The first five entries of the updated data frame are shown which contains the new columns with the personality subscales' scores.

In [11]:
personality_score_totals_df.head()

ID I am always prepared I am easily disturbed  \
0   0               (3, 5)                (4, 5)   
1   1               (3, 5)                (4, 5)   
2   2               (3, 5)                (4, 3)   
3   3               (3, 5)                (4, 5)   
4   4               (3, 3)                (4, 5)   

  I am exacting (demanding) in my work I am full of ideas  \
0                               (3, 5)             (5, 5)   
1                               (3, 5)             (5, 5)   
2                               (3, 3)             (5, 5)   
3                               (3, 3)             (5, 5)   
4                               (3, 3)             (5, 3)   

  I am interested in people I am not interested in abstract ideas  \
0                    (2, 3)                                (5, 3)   
1                    (2, 5)                                (5, 3)   
2                    (2, 5)                                (5, 5)   
3                    (2, 5)                                (5, 3)   
4                    (2, 3)                                (5, 3)   

  I am not interested in other people's problems  \
0                                         (2, 3)   
1                                         (2, 5)   
2                                         (2, 5)   
3                                         (2, 3)   
4                                         (2, 3)   

  I am not really interested in others I am quick to understand things  ...  \
0                               (2, 5)                          (5, 5)  ...   
1                               (2, 5)                          (5, 5)  ...   
2                               (2, 5)                          (5, 5)  ...   
3                               (2, 3)                          (5, 3)  ...   
4                               (2, 3)                          (5, 5)  ...   

  I sympathize with others' feelings I take time out for others  \
0                             (2, 5)                     (2, 5)   
1                             (2, 5)                     (2, 5)   
2                             (2, 5)                     (2, 5)   
3                             (2, 5)                     (2, 5)   
4                             (2, 3)                     (2, 5)   

  I talk to a lot of different people at parties I use difficult words  \
0                                         (1, 3)                (5, 1)   
1                                         (1, 5)                (5, 3)   
2                                         (1, 3)                (5, 1)   
3                                         (1, 5)                (5, 1)   
4                                         (1, 3)                (5, 1)   

  I worry about things Extraversion Agreeableness Conscientiousness  \
0               (4, 3)           30            40                48   
1               (4, 3)           42            46                46   
2               (4, 3)           28            40                40   
3               (4, 1)           30            38                38   
4               (4, 3)           28            34                46   

  Emotional stability Openness to experience  
0                  36                     42  
1                  40                     42  
2                  38                     42  
3                  40                     38  
4                  38                     36  

[5 rows x 56 columns]

The first five entries above indicate that the personality subscale scores and their respective columns have been added to the data frame.

The five newly added columns are "Conscientiousness", "Emotional stability", "Openness to experience", "Agreeableness'' and "Extraversion". These new columns are subscales for the questionnaire the applicants took as part of their application and evaluation.

Each of these columns were calculated by having the applicants' score responses to the questions categorized according to the subscales. With each response, the applicants provided a score as their response to the questions, which were stored as a tuple in the data set. The tuple had two values - the first value indicating which of the subscale the question belonged to and the second value in the tuple indicating the applicants' scored responses.

## Importing the departments dataset

The CSV file below is imported so that the dataset can be used to create the departments data frame. The departments data frame will then be used to merge it with the personality score totals data frame to create a merged data frame.

In [12]:
departments_df = pd.read_csv("../data/departments.csv", sep=";")

merged_personality_department_df = pd.concat(
    [
        personality_score_totals_df.reset_index(drop=True),
        departments_df["Department"].str.title().reset_index(drop=True),
    ],
    axis=1,
)

An assertion is made to determine if the newly created data frame has the expected shape and size. The newly created data frame should have the same number of rows as the "personality_score_totals_df" and it should have one column more than "personality_score_totals_df". The newly created data frame has one column more than "personality_score_totals_df" because the newly created data frame is the same as "personality_score_totals_df" with a column for the applicants' departments that they applied to being added to "personality_score_totals_df".

In [13]:
assert merged_personality_department_df.shape == (
    len(personality_score_totals_df),
    len(personality_score_totals_df.columns) + 1,
), "The dataframe does not have the expected shape - it has the incorrect number of rows and/or columns"

The first five entries in the merged personality and department data frame are displayed.

In [14]:
merged_personality_department_df.head()

ID I am always prepared I am easily disturbed  \
0   0               (3, 5)                (4, 5)   
1   1               (3, 5)                (4, 5)   
2   2               (3, 5)                (4, 3)   
3   3               (3, 5)                (4, 5)   
4   4               (3, 3)                (4, 5)   

  I am exacting (demanding) in my work I am full of ideas  \
0                               (3, 5)             (5, 5)   
1                               (3, 5)             (5, 5)   
2                               (3, 3)             (5, 5)   
3                               (3, 3)             (5, 5)   
4                               (3, 3)             (5, 3)   

  I am interested in people I am not interested in abstract ideas  \
0                    (2, 3)                                (5, 3)   
1                    (2, 5)                                (5, 3)   
2                    (2, 5)                                (5, 5)   
3                    (2, 5)                                (5, 3)   
4                    (2, 3)                                (5, 3)   

  I am not interested in other people's problems  \
0                                         (2, 3)   
1                                         (2, 5)   
2                                         (2, 5)   
3                                         (2, 3)   
4                                         (2, 3)   

  I am not really interested in others I am quick to understand things  ...  \
0                               (2, 5)                          (5, 5)  ...   
1                               (2, 5)                          (5, 5)  ...   
2                               (2, 5)                          (5, 5)  ...   
3                               (2, 3)                          (5, 3)  ...   
4                               (2, 3)                          (5, 5)  ...   

  I take time out for others I talk to a lot of different people at parties  \
0                     (2, 5)                                         (1, 3)   
1                     (2, 5)                                         (1, 5)   
2                     (2, 5)                                         (1, 3)   
3                     (2, 5)                                         (1, 5)   
4                     (2, 5)                                         (1, 3)   

  I use difficult words I worry about things Extraversion Agreeableness  \
0                (5, 1)               (4, 3)           30            40   
1                (5, 3)               (4, 3)           42            46   
2                (5, 1)               (4, 3)           28            40   
3                (5, 1)               (4, 1)           30            38   
4                (5, 1)               (4, 3)           28            34   

  Conscientiousness Emotional stability Openness to experience Department  
0                48                  36                     42       Data  
1                46                  40                     42       Data  
2                40                  38                     42       Data  
3                38                  40                     38       Data  
4                46                  38                     36       Data  

[5 rows x 57 columns]

Given that the assertions were true, the above displayed first five entries indicate that the data frames were successfully merged.

## Filtering the merged data frame

A risk status data frame is created from the newly merged data frame, which is filtered to determine which applicants are low and high risk based on their emotional stability, conscientiousness, and agreeableness subscale scores. The newly created risk status data frame is then updated to include a column for the applicants' risk status.

In [15]:
risk_status_df = merged_personality_department_df.copy()

for employee_id in range(len(merged_personality_department_df)):
    risk_status_df.at[employee_id, "Risk Status"] = np.where(
        (
            merged_personality_department_df["Emotional stability"].iloc[employee_id]
            < 30
            and merged_personality_department_df["Conscientiousness"].iloc[employee_id]
            < 30
            and merged_personality_department_df["Agreeableness"].iloc[employee_id] < 30
        ),
        "High risk",
        "Low risk",
    )

The first five entries in the new risk status data frame are displayed below.

In [16]:
risk_status_df.head()

ID I am always prepared I am easily disturbed  \
0   0               (3, 5)                (4, 5)   
1   1               (3, 5)                (4, 5)   
2   2               (3, 5)                (4, 3)   
3   3               (3, 5)                (4, 5)   
4   4               (3, 3)                (4, 5)   

  I am exacting (demanding) in my work I am full of ideas  \
0                               (3, 5)             (5, 5)   
1                               (3, 5)             (5, 5)   
2                               (3, 3)             (5, 5)   
3                               (3, 3)             (5, 5)   
4                               (3, 3)             (5, 3)   

  I am interested in people I am not interested in abstract ideas  \
0                    (2, 3)                                (5, 3)   
1                    (2, 5)                                (5, 3)   
2                    (2, 5)                                (5, 5)   
3                    (2, 5)                                (5, 3)   
4                    (2, 3)                                (5, 3)   

  I am not interested in other people's problems  \
0                                         (2, 3)   
1                                         (2, 5)   
2                                         (2, 5)   
3                                         (2, 3)   
4                                         (2, 3)   

  I am not really interested in others I am quick to understand things  ...  \
0                               (2, 5)                          (5, 5)  ...   
1                               (2, 5)                          (5, 5)  ...   
2                               (2, 5)                          (5, 5)  ...   
3                               (2, 3)                          (5, 3)  ...   
4                               (2, 3)                          (5, 5)  ...   

  I talk to a lot of different people at parties I use difficult words  \
0                                         (1, 3)                (5, 1)   
1                                         (1, 5)                (5, 3)   
2                                         (1, 3)                (5, 1)   
3                                         (1, 5)                (5, 1)   
4                                         (1, 3)                (5, 1)   

  I worry about things Extraversion Agreeableness Conscientiousness  \
0               (4, 3)           30            40                48   
1               (4, 3)           42            46                46   
2               (4, 3)           28            40                40   
3               (4, 1)           30            38                38   
4               (4, 3)           28            34                46   

  Emotional stability Openness to experience Department Risk Status  
0                  36                     42       Data    Low risk  
1                  40                     42       Data    Low risk  
2                  38                     42       Data    Low risk  
3                  40                     38       Data    Low risk  
4                  38                     36       Data    Low risk  

[5 rows x 58 columns]

The first five entries above indicate that the risk status for the respective applicants has been successfully calculated and that the data frame contains a new column for the respective risk statuses.

## Wrangling a new data frame

The respective high and low risk totals for the respective departments is calculated so that a risk status summary data frame can be created.

In [17]:
department_column_names_list = [
    "Risk Status",
    "Copywriting",
    "Data",
    "Design",
    "Strategy",
    "Web Dev",
]

risk_status_summary_df = pd.DataFrame(columns=department_column_names_list)

risk_status_summary_df["Risk Status"] = ["Low Risk", "High Risk"]


def risk_score_calculator(department):
    risk_grouped_by_department_name = risk_status_df.groupby("Department").get_group(
        department
    )["Risk Status"]

    low_risk_score = 0
    high_risk_score = 0

    for risk_status in risk_grouped_by_department_name:
        if risk_status == "Low risk":
            low_risk_score += 1

        else:
            high_risk_score += 1

    risk_status_summary_df[department] = [low_risk_score, high_risk_score]


for department in department_column_names_list[1:]:
    risk_score_calculator(department)

The risk status summary totals are displayed for the respective departments.

In [18]:
display(risk_status_summary_df.style.hide())

With the risk status table, each department was analyzed for potential high and low risk applicants. This was done by checking which applicants had a total score of less than 30 in the "Agreeableness", "Conscientiousness", and "Emotional stability" categories. If an applicant has a score of less than 30 in all three respective categories, then they are deemed "High risk".

Only 2 applicants are high risk based on the risk status summary. This indicates that 0.12% of the applicants are high risk while 99.88% of the applicants are low risk.

# Conclusion 

Using the various datasets, methods and calculations, certain outcomes and values can be determined based on conditions and the personality scores of the respective applicants to decide the outcome of their applications.

From the calculations made based on the applicants' personality scores, only 2 applicants are high risk based on the risk status summary. This indicates that 0.12% of the applicants are high risk while 99.88% of the applicants are low risk based on the personality scores they took as part of their application. 

As such, this indicates that there is an overwhelming number of applicants who are low risk while the high risk applicants only represent a minute number of the applicants. 